translator/trace/zipkin/protospan_to_zipkinv1.go

```diff
@@ -0,0 +1,323 @@
+// zipkin endpoint and insert the attribute keys that were made redundant
+// (because now they can be represented by the endpoint) into the redundantKeys
+// map (function assumes that this was created by the caller). Per call at most
+// 3 attribute keys can be made redundant.
+func zipkinEndpointFromAttributes(
+	attributes map[string]interface{},
+	serviceName string,
+	endpointType zipkinDirection,
+	redundantKeys map[string]bool,
+) (endpoint *zipkinmodel.Endpoint) {
+
+	if attributes == nil {
+		return nil
+	}
+
+	// The data in the Attributes map was saved in the format
+	// {
+	//      "ipv4": "192.168.99.101",
+	//      "port": "9000",
+	//      "serviceName": "backend",
+	// }
+
+	var ipv4Key, ipv6Key, portKey string
+	if endpointType == isLocalEndpoint {
+		ipv4Key, ipv6Key, portKey = LocalEndpointIPv4, LocalEndpointIPv6, LocalEndpointPort
+	} else {
+		ipv4Key, ipv6Key, portKey = RemoteEndpointIPv4, RemoteEndpointIPv6, RemoteEndpointPort
+	}
+
+	var ip net.IP
+	ipv6Selected := false
+
+	if ipv4Str, ok := extractStringAttribute(attributes, ipv4Key); ok {
+		ip = net.ParseIP(ipv4Str)
+		redundantKeys[ipv4Key] = true
+	} else if ipv6Str, ok := extractStringAttribute(attributes, ipv6Key); ok {
+		ip = net.ParseIP(ipv6Str)
+		ipv6Selected = true
+		redundantKeys[ipv6Key] = true
+	}
+
+	var port uint64
+	if portStr, ok := extractStringAttribute(attributes, portKey); ok {
+		port, _ = strconv.ParseUint(portStr, 10, 16)
+		redundantKeys[portKey] = true
+	}
+
+	if serviceName == "" && len(ip) == 0 && port == 0 {
+		// Nothing to put on the endpoint
+		return nil
+	}
+
+	zEndpoint := &zipkinmodel.Endpoint{
+		ServiceName: serviceName,
+		Port:        uint16(port),
+	}
+
+	if ipv6Selected {
+		zEndpoint.IPv6 = ip
+	} else {
+		zEndpoint.IPv4 = ip
+	}
+
+	return zEndpoint
+}
+
+func extractStringAttribute(
+	attributes map[string]interface{},
+	key string,
+) (value string, ok bool) {
+	var i interface{}
+	if i, ok = attributes[key]; ok {
+		value, ok = i.(string)
+	}
+
+	return value, ok
+}
+
+func convertTraceID(t trace.TraceID) zipkinmodel.TraceID {
+	h, l, _ := tracetranslator.BytesToUInt64TraceID(t[:])
+	return zipkinmodel.TraceID{High: h, Low: l}
+}
+
+func convertSpanID(s trace.SpanID) zipkinmodel.ID {
+	id, _ := tracetranslator.BytesToUInt64SpanID(s[:])
+	return zipkinmodel.ID(id)
+}
+
+func spanKind(s *trace.SpanData) zipkinmodel.Kind {
+	// First try span kinds that are set directly in SpanKind field.
+	switch s.SpanKind {
+	case trace.SpanKindClient:
+		return zipkinmodel.Client
+	case trace.SpanKindServer:
+		return zipkinmodel.Server
+	}
+
+	// SpanKind==SpanKindUnspecified, check if TagSpanKind attribute is set.
+	// This can happen if span kind had no equivalent in OC, so we could represent it in
+	// the SpanKind. In that case we had set a special attribute TagSpanKind in
+	// the span to preserve the span kind. Now we will do a reverse translation.
+	spanKind, ok := s.Attributes[tracetranslator.TagSpanKind]
```

Comment:
side note: this is one of those cases where nil has some behavior, in this case it returns the zero value of the respective map value and !ok.

Reply:
Changed to make it more apparent that we will not have a nil value.